src/archive.rs

```diff
@@ -316,234 +236,113 @@ impl<'a> Archive<Read + 'a> {
-    ///
-    /// # Errors
-    ///
-    /// This function will return an error for any intermittent I/O error which
-    /// occurs when either reading or writing.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use tar::{Archive, Header};
-    ///
-    /// let mut header = Header::new();
-    /// header.set_path("foo");
-    /// header.set_size(4);
-    /// header.set_cksum();
-    ///
-    /// let mut data: &[u8] = &[1, 2, 3, 4];
-    ///
-    /// let mut ar = Archive::new(Vec::new());
-    /// ar.append(&header, &mut data).unwrap();
-    /// let archive = ar.into_inner();
-    /// ```
-    pub fn append(&self, header: &Header, data: &mut Read) -> io::Result<()> {
-        let me: &Archive<Write> = self;
-        me._append(header, data)
-    }
+    fn next_entry(&mut self) -> io::Result<Option<Entry<'a, io::Empty>>> {
+        if self.raw {
+            return self.next_entry_raw()
+        }
 
-    /// Adds a file on the local filesystem to this archive.
-    ///
-    /// This function will open the file specified by `path` and insert the file
-    /// into the archive with the appropriate metadata set, returning any I/O
-    /// error which occurs while writing. The path name for the file inside of
-    /// this archive will be the same as `path`, and it is recommended that the
-    /// path is a relative path.
-    ///
-    /// Note that this will not attempt to seek the archive to a valid position,
-    /// so if the archive is in the middle of a read or some other similar
-    /// operation then this may corrupt the archive.
-    ///
-    /// Also note that after all files have been written to an archive the
-    /// `finish` function needs to be called to finish writing the archive.
-    ///
-    /// # Examples
-    ///
-    /// ```no_run
-    /// use tar::Archive;
-    ///
-    /// let mut ar = Archive::new(Vec::new());
-    ///
-    /// ar.append_path("foo/bar.txt").unwrap();
-    /// ```
-    pub fn append_path<P: AsRef<Path>>(&self, path: P) -> io::Result<()> {
-        let me: &Archive<Write> = self;
-        me._append_path(path.as_ref())
-    }
+        let mut gnu_longname = None;
+        let mut gnu_longlink = None;
+        let mut pax_extensions = None;
+        let mut processed = 0;
+
+        loop {
+            processed += 1;
+            let entry = match try!(self.next_entry_raw()) {
+                Some(entry) => entry,
+                None if processed > 1 => {
+                    return Err(other("members found describing a future member \
+                                      but no future member found"))
+                }
+                None => return Ok(None),
+            };
+
+            if entry.header().as_gnu().is_some() &&
+               entry.header().entry_type().is_gnu_longname() {
+                if gnu_longname.is_some() {
+                    return Err(other("two long name entries describing \
+                                      the same member"))
```

Comment:
I'm not actually sure, I couldn't necessarily find documentation indicating one way or another. I figured it'd be good to flag and be explicit, though, to prevent silently ignoring any data.